### PR TITLE
docs: render options with role

### PR DESCRIPTION
**Pull request by tchaikov**
this series tries to 

1. render options with role. so the options can be cross referenced and defined.
2. move the formatting out of the content. so the representation can be defined in a more flexible way.

### CODE REVIEW

**Comment by tzach** [comment]
@dgarcia360 please take a look

**Comment by tzach** [comment]
@tchaikov to validate: this PR refactors how configuration parameters style is generated during doc make.
It keeps the documentation feel and look the same.

**Comment by tchaikov** [comment]
@tzach Hi Tzach, yes, i intentionally keep the rendering the same in this change. will leave the change of the representation to follow-up PRs.

**Comment by tchaikov** [comment]
> The page is a bit different now.
> 
>     * The horizontal line added with [this PR](https://github.com/scylladb/scylladb/pull/15827) is missing (but it may be OK, as the headings look better now).
> 
>     * The headings are different:
>       
>       * They look different (better).
>       * They do not constitute links, which is a critical thing. In the current documentation, the user can easily get the link to share the parameter description.
> 
> 
> image
> 
> This is what I can review, but this PR must be reviewed by @dgarcia360 before it is merged.

@annastuchlik could you please update the local cache and try again? the changes you noticed are all in reflected in the updated CSS.

**Comment by tchaikov** [comment]
image

this is what i have locally.

**Comment by dgarcia360** [comment]
I'd like to discuss a few points regarding the changes:

* **Separation of formatting and parsing**: I've seen you deleted the Jinja template, and mimick the behaviour with Python instead. My first impresion is that it does not provide more flexibility, but the opposite. Now, we generate a JSON object with the flexiblity to be rendered differently via a template. Now, if you want to change the look and feel, you have to do changing the Python code. 

If the goal is better support for cross-references, wouldn't it be more straightforward to edit the template directly?

* **Cross-referencing options with roles**: 

I see two ways we could achieve this:

1. Edit the template so that it includes a role before rendering the heading:

```
(...)
{% for item in group.properties %}
{% if item.value_status == value_status %}
.. _config_{{item.name:
{{ item.name }}
```



I'm curious about the specific benefits this change brings. Could you shed more light on this?




* **Cross-referencing options with roles:** The current setup already lets you cross reference options. For example:

  ```
  
  ```
  
  Could you provide more insight into the specific advantages of this change?

**Comment by dgarcia360** [comment]
I'd like to discuss a few points regarding the changes:

* **Formatting vs. Parsing**: : I noticed the Jinja template was removed in favor of defining the UI with Python. Initially, it seems this might limit flexibility rather than enhancing it. Now, we have have a separation of concerns:

- **Parser:** Generates a JSON object that could be rendered variably via a template. 
- **Template:** Defines how to render the JSON object.

With the proposed changes, adjustments to the appearance would require modifying the Python code. 

* **Cross-referencing options with roles**: 

To achieve this, while retaining the separation between rendering and parsing, we could just update the template to include a role before the heading:

```python
(...)
{% for item in group.properties %}
{% if item.value_status == value_status %}
.. _config_{{item.name}} # --> add this line

{{ item.name }}
(...)
```

This should allow us to cross reference the properties as follows:

```rst
:ref:`Link to sample config param <config_ memtable_flush_static_shares >`.
```

**Comment by tchaikov** [comment]
> I'd like to discuss a few points regarding the changes:
> 
>     * **Formatting vs. Parsing**: : I noticed the Jinja template was removed in favor of defining the UI with Python. Initially, it seems this might limit flexibility rather than enhancing it. Now, we have have a separation of concerns:
> 

the jinja template was moved from a separated file into the Python code. so the flexibility is not damaged. if we prefer putting the jinja template into a separated file, we can. but since the template references the values with the key defined by the dict offered by python code, and it's not super complicated, why don't we colocate them? 

> 
>     * **Parser:** Generates a JSON object that could be rendered variably via a template.
> 
>     * **Template:** Defines how to render the JSON object.
> 

this still holds. to be more specific, the parser returns a dict, which can be consumed by jinja templates.

> 
> With the proposed changes, adjustments to the appearance would require modifying the Python code.

explained above. the jinja templates are embedded in python. if we consider it as Python code, well, yes.

> 
>     * **Cross-referencing options with roles**:
> 
> 
> To achieve this, while retaining the separation between rendering and parsing, we could just update the template to include a role before the heading:
> 
> ```python
> (...)
> {% for item in group.properties %}
> {% if item.value_status == value_status %}
> .. _config_{{item.name}} # --> add this line
> 
> {{ item.name }}
> (...)
> ```
> 
> This should allow us to cross reference the properties as follows:
> 

yeah, we can always define a hyperlink this way. but i would prefer using the sphinx way to define objects and to reference them, and render them, instead of flattening these structured information into HTML. and by structuring the options as object offers better readability. the reason i prefer using CSS instead of raw HTML is in the same spirit.

> ```
> :ref:`Link to sample config param <config_ memtable_flush_static_shares >`.
> ```

**Comment by dgarcia360** [comment]
Thanks for explaining! I see your points. 

> if we prefer putting the jinja template into a separated file, we can. but since the template references the values with the key defined by the dict offered by python code, and it's not super complicated, why don't we colocate them?

It's a personal preference (not a blocker). Having the templates in different files:
-  Facilitates contributors unfamiliar with Python to edit how the content will render.
- Lets define with less effort new templates to generate the same content differently. 
- Matches how we autogenerate docs in other projects via custom extensions.

By the way, I tested it locally, and everything looks great!

**Comment by tchaikov** [comment]
> Thanks for explaining! I see your points.
> 
> > if we prefer putting the jinja template into a separated file, we can. but since the template references the values with the key defined by the dict offered by python code, and it's not super complicated, why don't we colocate them?
> 
> It's a personal preference (not a blocker). Having the templates in different files:
> 
>     * Facilitates contributors unfamiliar with Python to edit how the content will render.
> 
>     * Lets define with less effort new templates to generate the same content differently.
> 
>     * Matches how we autogenerate docs in other projects via custom extensions.
> 
> 
> By the way, I tested it locally, and everything looks great!

i see. will put the templates back into separated files in the next revision.

**Comment by tchaikov** [comment]
@dgarcia360 hi David, updated and repushed. i think the next step would be to use field list to define the properties of an option, so we can drop the code for rendering in `ConfigOption`, and just define the field list in `db_config.tmpl`. there are some pending works in my mind:

- to format the description using ReST. i think this might be a little bit difficult. as we kind of format the description in free style =)
- to customize the field list, so it can keep the existing layout, which looks nice. this part is trivial, but not that difficult, i guess.

anyway, i think this changeset has achieved its goals: 

1. to drop the yaml file
2. to define the options with role
3. to further split the formatting and content

File: docs/_ext/scylladb_cc_properties.py [hunk 1]
```diff
@@ -1,14 +1,23 @@
...
+from sphinx.jinja2glue import BuiltinTemplateLoader
+from docutils import nodes
 from docutils.parsers.rst import directives
+from docutils.statemachine import StringList
 
-CONFIG_FILE_PATH = "../db/config.cc"
-CONFIG_HEADER_FILE_PATH = "../db/config.hh"
-DESTINATION_PATH = "_data/db_config.yaml"
-
+logger = logging.getLogger(__name__)
```
Comment (tchaikov) [inline docs/_ext/scylladb_cc_properties.py:20]: it's added just for debugging.

**Comment by dgarcia360** [comment]
LGTM

> to format the description using ReST. i think this might be a little bit difficult. as we kind of format the description in free style =)

I'm happy to submit a pull request to adapt `config.cc` descriptions in rest format. By doing so, we could remove or simpify the `readable_desc` function.

**Comment by tchaikov** [comment]
v3:

- rebased against master HEAD

**Comment by tchaikov** [comment]
> LGTM
> 
> > to format the description using ReST. i think this might be a little bit difficult. as we kind of format the description in free style =)
> 
> I'm happy to submit a pull request to adapt `config.cc` descriptions in RST format. By doing so, we could remove or simpify the `readable_desc` function.

great! i am all for this. also, i think the RST should looks good even printed directly in console, so guess it won't be controversial change.

**Comment by tchaikov** [comment]
@annastuchlik yeah, i think so.

@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Comment by tchaikov** [comment]
> @tchaikov I think this PR must be backported to branch-5.4. Is that correct? The Parameters page is working on master: https://opensource.docs.scylladb.com/master/reference/configuration-parameters.html The list parameters is no longer generated on branch-5.4 after this PR was merged: https://opensource.docs.scylladb.com/branch-5.4/reference/configuration-parameters.html

@annastuchlik hi Anna, i don't think so. as it is not considered as fix or is intrusive to other branches. regarding the missing parameter list in branch-5.4. i just rebuilt the docs using 

```console
$ git worktree add branch-5.4 branch-5.4
$ cd branch-5.4/docs
$ make setup
$ make preview
```

and then point my browser to http://127.0.0.1:5500/reference/configuration-parameters/ . i can see the parameter list. may i learn how we generate the document for branch-5.4? can we audit the build process to understand what was going wrong?

**Comment by dgarcia360** [comment]
@annastuchlik @tchaikov We need to backport this to `branch-5.4`, since the name of the directive changed in https://github.com/scylladb/scylladb/pull/15860/files#diff-2f8f519584941f4147ad3f144e95fe2fd9d8dbfd898830ed53a32dce1d588ba8R8.

When building the docs with the multiversion option, it loads the latest version of the extension (available in `master`) to generate the docs for the different versions. This isn't visible for branch-5.6 since it does not recognize the new directive name.

**Comment by tchaikov** [comment]
@annastuchlik hi Anna, thank you for the explanations! I did try the multiversion task , but it failed due to some Java related issues. Please go in and request the backport from our maintainers.

**Comment by denesb** [comment]
Backport is not clean, need a backport PR.

**Comment by tchaikov** [comment]
ack. lemme prepare it.

**Comment by tchaikov** [comment]
@annastuchlik thank you for the confirmation! it works!

**Pull request closed without merging**